### PR TITLE
fix(rest): create a new endpoint to save fossology connection in admin tab.

### PR DESCRIPTION
**Pull request by nikkuma7**
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? (*Refer to issue here*)
> * Did you add or update any new dependencies that are required for your change?

Issue: #2198 

### How To Test?
http://localhost:8080/resource/api/fossology/saveConfig

> How should these changes be tested by the reviewer?
> Have you implemented any additional tests?

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

### CODE REVIEW

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/FossologyAdminController.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    Sw360FossologyAdminServices sw360FossologyAdminServices;
+
+    @Override
+    public RepositoryLinksResource process(RepositoryLinksResource resource) {
+        resource.add(linkTo(FossologyAdminController.class).slash("api" + FOSSOLOGY_URL).withRel("fossology"));
+        return resource;
+    }
+
+    @PostMapping(value = FOSSOLOGY_URL + "/saveConfig", consumes  = {MediaType.APPLICATION_JSON_VALUE})
+    public ResponseEntity<?> saveConfigration(@RequestBody Map<String, String> request) throws TException {
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/FossologyAdminController.java:72]: typo saveConfigration -> saveConfiguration

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/FossologyAdminController.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.sw360.rest.resourceserver.admin.fossology;
+
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
+
+import java.util.Map;
+import java.util.stream.Collectors;
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/FossologyAdminController.java:17]: unused imports: line 17-20, 23-31, 35, 38, 40, 46-50

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/Sw360FossologyAdminServices.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.Set;
+
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.THttpClient;
+import org.apache.thrift.transport.TTransportException;
+import org.eclipse.sw360.datahandler.permissions.PermissionUtils;
+import org.eclipse.sw360.datahandler.thrift.ConfigContainer;
+import org.eclipse.sw360.datahandler.thrift.RequestStatus;
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/Sw360FossologyAdminServices.java:25]: unused imports: line 25-27, 33

File: rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/fossologySpecTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.hateoas.MediaTypes;
+import org.springframework.http.MediaType;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+public class fossologySpecTest extends TestRestDocsSpecBase {
```
Comment (hoangnt2) [inline rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/fossologySpecTest.java:44]: Class names should start with a capital letter: rename file fossologySpecTest -> FossologySpecTest

**Comment by hoangnt2** [comment]
Test successfully

image

image

**Comment by hoangnt2** [comment]
@nikkuma7, I have a question. I used API to save fossology configuration. It worked. But after that I shutdown server and restart it, I couldn't access liferay GUI. 
Did you face this problem from your side

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/FossologyAdminController.java (hunk 1 above)
Comment (nikkuma7) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/FossologyAdminController.java:56]: I have already renamed this method in PR: 2212

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/FossologyAdminController.java (hunk 2 above)
Comment (nikkuma7) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/FossologyAdminController.java]: done.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/Sw360FossologyAdminServices.java (hunk 3 above)
Comment (nikkuma7) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/admin/fossology/Sw360FossologyAdminServices.java:25]: done.

File: rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/fossologySpecTest.java (hunk 4 above)
Comment (nikkuma7) [reply rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/fossologySpecTest.java]: done.

**Comment by nikkuma7** [comment]
> @nikkuma7, I have a question. I used API to save fossology configuration. It worked. But after that I shutdown server and restart it, I couldn't access liferay GUI. Did you face this problem from your side

@hoangnt2 
Below steps follow:
1: create a fossology token.
2: save configuration.
3: trigger re check connection (get success).
4: shut down the server.
5: restart the server and check the GUI (showing success).

I shut down server After save fossology configuration. GUI is showing success.

image
image

**Comment by hoangnt2** [comment]
@nikkuma7, I tried your case, it works. 
But when I change the URL, replace the correct IP address by an IP address that doesn't exist. Then I stopped and restarted server, I couldn't open GUI anymore.

**Review by hoangnt2 — COMMENTED** [review]
Code look good to me

**Pull request merged**